### ISSUE TITLE
Another example of slow ODEProblem and solve

### ISSUE THREAD

**Issue by jaakkor2** [SciML/ModelingToolkit.jl#1962]
This is an example what looks a bit like https://github.com/SciML/ModelingToolkit.jl/issues/1285, but is more simple.  https://github.com/SciML/ModelingToolkit.jl/issues/1286 and https://github.com/SciML/ModelingToolkit.jl/issues/1461 could be related.

```
using OrdinaryDiffEq
using ModelingToolkit

n, m = 60, 12
ω = range(1, m, length=m)
K = ω.^2
D = ω / 10
A = 0.1 * rand(n,m)
B = 0.1 * rand(n,m)
du0 = zeros(m)
u0 = zeros(m)
u0[1] = 1
tspan = (0, 10)
saveat = 1/100

function f(ddu,du,u,p,t)
    F = - B' * 1e-2*((1 .+ A*u) ./ (1 .+ B*u).^2)
    ddu .= -K .* u - D .* du + F
    nothing
end

prob = SecondOrderODEProblem(f,du0,u0,tspan)

println("Times for ODE")
@time sol = solve(prob, Tsit5(), saveat=saveat)
@time sol = solve(prob, Tsit5(), saveat=saveat)

println("Times for MTK")
@time sys = modelingtoolkitize(prob)
#@time sys = structural_simplify(sys)
@time prob_MTK = ODEProblem(sys, sparse=false, jac=true, simplify=false)
@time sol_MTK = solve(prob_MTK, Tsit5(), saveat=saveat)
@time sol_MTK = solve(prob_MTK, Tsit5(), saveat=saveat)

nothing
```
With Julia 1.8.3,  OrdinaryDiffEq v6.31.2, ModelingToolkit v8.34.0, I get
```
Times for ODE
  3.971011 seconds (7.65 M allocations: 363.034 MiB, 3.33% gc time, 99.88% compilation time: 100% of which was recompilation)
  0.004545 seconds (18.03 k allocations: 7.483 MiB)
Times for MTK
  1.095540 seconds (12.01 M allocations: 563.597 MiB, 9.82% gc time, 17.06% compilation time: 100% of which was recompilation)
 30.252890 seconds (259.41 M allocations: 12.328 GiB, 14.56% gc time, 2.46% compilation time)
 10.518277 seconds (36.68 M allocations: 1.756 GiB, 7.83% gc time, 99.93% compilation time)
  0.005795 seconds (1.05 k allocations: 308.781 KiB)
```
i.e., MTK is 10x slower than ODE for the first run.

The ratio gets worse if parameter values of `n` and `m` are increased.  With `n, m = 120, 24` I get
```
Times for ODE
  3.763278 seconds (7.66 M allocations: 393.806 MiB, 99.51% compilation time: 100% of which was recompilation)
  0.023134 seconds (28.20 k allocations: 38.255 MiB)
Times for MTK
  8.241956 seconds (92.73 M allocations: 4.132 GiB, 15.12% gc time, 2.39% compilation time: 94% of which was recompilation)
674.101274 seconds (3.83 G allocations: 183.205 GiB, 12.05% gc time, 0.02% compilation time)
142.015902 seconds (249.25 M allocations: 10.127 GiB, 4.40% gc time, 99.95% compilation time)
  0.073260 seconds (1.05 k allocations: 499.656 KiB)
```
i.e., now MTK is 200x slower than ODE.

**Comment by ChrisRackauckas** [comment]
Yes, this is exactly the same as some of the other issues on this.

**Comment by jaakkor2** [comment]
Testing with 1.9.0-alpha1, I get for `n, m = 60, 12`
```
Times for ODE
  2.266930 seconds (2.31 M allocations: 122.160 MiB, 99.78% compilation time: 100% of which was recompilation)
  0.004910 seconds (18.61 k allocations: 7.347 MiB)
Times for MTK
  3.020208 seconds (14.91 M allocations: 699.476 MiB, 14.35% gc time, 1.32% compilation time: 100% of which was recompilation)
 48.601693 seconds (359.22 M allocations: 17.241 GiB, 13.95% gc time, 0.17% compilation time)
  7.735031 seconds (10.55 M allocations: 554.426 MiB, 99.90% compilation time)
  0.007494 seconds (1.05 k allocations: 306.516 KiB)
```
i.e., ODE improves, MTK degrades.

**Comment by YingboMa** [comment]
Close for duplication. You are mostly just comparing compilation time not true solve time. There probably shouldn't be a case where you run problem construction in a loop.

**Comment by jaakkor2** [comment]
> You are mostly just comparing compilation time not true solve time.

All my timing numbers are for the second time the snippet above is run.  Again with Julia 1.9.0-alpha1, `n, m = 60, 12`

First run
```
Times for ODE
  7.080068 seconds (17.37 M allocations: 1.048 GiB, 4.08% gc time, 99.88% compilation time)
  0.044771 seconds (19.79 k allocations: 7.478 MiB, 87.72% gc time)
Times for MTK
 18.340798 seconds (46.56 M allocations: 2.890 GiB, 3.35% gc time, 86.55% compilation time)
 78.541151 seconds (426.81 M allocations: 21.814 GiB, 9.53% gc time, 40.00% compilation time)
  7.180618 seconds (10.61 M allocations: 554.627 MiB, 2.24% gc time, 99.89% compilation time)
  0.007457 seconds (1.05 k allocations: 306.516 KiB)
```
Second run
```
Times for ODE
  2.298641 seconds (2.31 M allocations: 122.377 MiB, 99.80% compilation time: 100% of which was recompilation)
  0.004999 seconds (20.01 k allocations: 7.565 MiB)
Times for MTK
  2.672795 seconds (14.74 M allocations: 692.066 MiB, 6.61% gc time, 3.07% compilation time: 46% of which was recompilation)
 47.710714 seconds (358.17 M allocations: 17.184 GiB, 12.83% gc time, 0.17% compilation time)
  7.418660 seconds (10.55 M allocations: 551.423 MiB, 2.98% gc time, 99.89% compilation time)
  0.007669 seconds (1.05 k allocations: 306.516 KiB)
```
So the solve time for the same problem for ODE is 2 s and for MTK 48 s. The time to solve the _same_ problem the second time is the same about 7 ms.

And this is for a very small test case, realistic models are much larger, and above it was shown that MTK problem construction time scales very badly with problem size.

> There probably shouldn't be a case where you run problem construction in a loop.

Maybe I am misunderstanding something very basic, but if one wants to solve a problem, it needs to be constructed with `ODEProblem`.  And then solving for another problem, `ODEProblem` needs to be called again?

Closing the issue is fine, but judging from the status of https://github.com/SciML/ModelingToolkit.jl/issues/1285, I need to keep using ODE, and not even consider MTK ?

**Comment by ChrisRackauckas** [comment]
> Maybe I am misunderstanding something very basic, but if one wants to solve a problem, it needs to be constructed with ODEProblem. And then solving for another problem, ODEProblem needs to be called again?

No, you just use `remake`. You should just compile only once for things like inverse problems.

> Closing the issue is fine, but judging from the status of https://github.com/SciML/ModelingToolkit.jl/issues/1285, I need to keep using ODE, and not even consider MTK ?

It depends on what you're doing, and what your time frame is. MTK's compilation path is being overhauled to retain more structure and thus decrease compilation cost on models with structure. In cases like this, it should in the future be able to compile in O(1) time, though right now it's O(nm). That's being worked on, and you can follow the progress in the repo, but it's not here today, and it's a duplicate of other discussions about this topic. If you need something today, then stick with DifferentialEquations.jl direct and just write the function. Meanwhile, MTK is continuing to improve on this front and it's rather binary: once it changes from O(nm) to O(1) the representation is dramatically different and the real issue is fixed. But because it's a very binary change, little details of v1.9 vs v1.8 just aren't relevant or useful: see the other issues, it's something that's known and not something that can be chipped away at as its instead something that needs a different computational complexity.

> And this is for a very small test case, realistic models are much larger

Of course, this is just silly. "Realistic" means many things. What you mean here is small models with lots of structure. That's MTK's worst nightmare right now, so for now it needs a lot of work to handle that better in terms of compilation. Large models which are relatively unstructured, like chemical reaction networks and acausal models without embedded PDEs, are where MTK currently shines because there isn't much structure to retain and thus the compilation time matches what "a better person would have written by hand" (post tearing and all). So what you really mean here is that, if there is a repeated structure and that structure is large, ModelingToolkit's code generation will generate an amount of code that is sized O(n) to match the size of the structure, instead of being O(1) and retaining the control flow (loops, matrix multiplications), and thus changing dimensional parameters changes code size and compile time. This of course is a known consequence of today's code generation, and is being worked on, but is the reason why a lot of these applications (like PDE discretizations) are still demonstrated with pure DifferentialEquations.jl. And it's fine! We write both libraries, we're okay with people not using MTK for that since it's not optimized there right now!

**Comment by jaakkor2** [comment]
Thank you @ChrisRackauckas for the clarification of the state of things.  Good to know that the issue is well known (but was a bit of a surprise for me) and will hopefully be improved in the future.
